Sources/Source/SourceLocation.swift

```diff
@@ -55,3 +61,17 @@ extension SourceRange: Equatable {
     return lhs.start == rhs.start && lhs.end == rhs.end
   }
 }
+
+extension SourceRange {
+  private var contents: String { return start.file.contents }
+
+  public var length: Int {
+    return end.charOffset - start.charOffset
+  }
+
+  public var source: String {
+    let startIndex = contents.index(contents.startIndex, offsetBy: start.charOffset)
+    let endIndex = contents.index(contents.startIndex, offsetBy: end.charOffset)
```

Comment:
pass startIndex as the first argument and use the difference of the offsets as the offset from startIndex to the end. That should save a bunch of work.